diff --git a/config/hal.c b/config/hal.c
index 3479be9..a235876 100644
--- a/config/hal.c
+++ b/config/hal.c
@@ -177,7 +177,13 @@ device_added(LibHalContext *hal_ctx, const char *udi)
         if (strcmp(props[i], "input.keys") == 0 ||
             strcmp(props[i], "input.keyboard") == 0)
             type |= TYPE_KEYS;
         if (strcmp(props[i], "input.mouse") == 0)
+        {
+            type |= TYPE_POINTER;
+            add_option(&options, "Mode", "Relative");
+        } else if (strcmp(props[i], "input.touchpad") == 0) {
             type |= TYPE_POINTER;
+            add_option(&options, "Mode", "Absolute");
+        }
     }
     libhal_free_string_array(props);
diff --git a/config/x11-input.fdi b/config/x11-input.fdi
index c390706..20c73fa 100644
--- a/config/x11-input.fdi
+++ b/config/x11-input.fdi
@@ -1,7 +1,14 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <deviceinfo version="0.2">
   <device>
-    <!-- FIXME: Support tablets too. -->
+    <match key="info.capabilities" contains="input.touchpad">
+      <merge key="input.x11_driver" type="string">mouse</merge>
+      <match key="/org/freedesktop/Hal/devices/computer:system.kernel.name"
+             string="Linux">
+        <merge key="input.x11_driver" type="string">evdev</merge>
+      </match>
+    </match>
+
     <match key="info.capabilities" contains="input.mouse">
       <merge key="input.x11_driver" type="string">mouse</merge>
       <match key="/org/freedesktop/Hal/devices/computer:system.kernel.name"
@@ -16,6 +23,25 @@
       <!-- If we're using Linux, we use evdev by default (falling back to
            keyboard otherwise). -->
       <merge key="input.x11_driver" type="string">keyboard</merge>
+      <merge key="input.xkb.model" type="string">olpc</merge>
+      <match key="/org/freedesktop/Hal/devices/computer:system.kernel.name"
+             string="Linux">
+        <merge key="input.x11_driver" type="string">evdev</merge>
+        <merge key="input.xkb.model" type="string">evdev</merge>
+      </match>
+
+      <merge key="input.xkb.layout" type="string">us</merge>
+
+      <merge key="input.xkb.variant" type="string" />
+    </match>
+
+    <!-- older versions of HAL use "keyboard" instead of keys -->
+    <match key="info.capabilities" contains="input.keyboard">
+      <merge key="input.xkb.rules" type="string">base</merge>
+
+      <!-- If we're using Linux, we use evdev by default (falling back to
+           keyboard otherwise). -->
+      <merge key="input.x11_driver" type="string">keyboard</merge>
       <merge key="input.xkb.model" type="string">pc105</merge>
       <match key="/org/freedesktop/Hal/devices/computer:system.kernel.name"
              string="Linux">
diff -u -r a/config/hal.c a/config/hal.c
--- a/config/hal.c	2008-03-23 10:43:43.000000000 -0400
+++ b/config/hal.c	2008-03-23 10:41:17.000000000 -0400
@@ -163,6 +163,14 @@
 
     dbus_error_init(&error);
 
+    options = xcalloc(sizeof(*options), 1);
+    options->key = xstrdup("_source");
+    options->value = xstrdup("server/hal");
+    if (!options->key || !options->value) {
+        ErrorF("[config] couldn't allocate first key/value pair\n");
+        goto unwind;
+    }
+
     props = libhal_device_get_property_strlist(hal_ctx, udi,
                                                "info.capabilities", &error);
     if (!props) {
@@ -209,14 +217,6 @@
         xkb_options = get_prop_string_array(hal_ctx, udi, "input.xkb.options");
     }
 
-    options = xcalloc(sizeof(*options), 1);
-    options->key = xstrdup("_source");
-    options->value = xstrdup("server/hal");
-    if (!options->key || !options->value) {
-        ErrorF("[config] couldn't allocate first key/value pair\n");
-        goto unwind;
-    }
-
     add_option(&options, "path", path);
     add_option(&options, "driver", driver);
     add_option(&options, "name", name);
