From c45a1c190b06ae7f840cd341d7865224bb95fca0 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Wed, 21 Oct 2009 15:55:56 -0400
Subject: [PATCH] fix build with new automake

---
diff --git a/hw/xfree86/common/Makefile.am b/hw/xfree86/common/Makefile.am
index 241736b..2a610ae 100644
--- a/hw/xfree86/common/Makefile.am
+++ b/hw/xfree86/common/Makefile.am
@@ -11,7 +11,6 @@ endif
 
 if XV
 XVSOURCES = xf86xv.c xf86xvmc.c
-XVSDKINCS = xf86xv.h xf86xvmc.h
 endif
 
 XKBSOURCES = xf86XKB.c
diff --git a/hw/xfree86/parser/Makefile.am b/hw/xfree86/parser/Makefile.am
index 0691158..a0158b8 100644
--- a/hw/xfree86/parser/Makefile.am
+++ b/hw/xfree86/parser/Makefile.am
@@ -1,8 +1,5 @@
 if INSTALL_LIBXF86CONFIG
 lib_LIBRARIES = libxf86config.a
-LIBHEADERS = \
-	xf86Optrec.h \
-	xf86Parser.h
 else
 noinst_LIBRARIES = libxf86config.a
 endif
-- 
1.6.4.4

